File: java/bc-api-service/src/main/kotlin/io/mike/api/contracts/books/BookService.kt
package io.mike.api.contracts.books

import io.mike.api.contracts.books.filters.applyBookSearchFilter
import io.mike.api.contracts.books.support.BookExtendedResponse
import io.mike.foundation.repository.PageResponse
import io.mike.foundation.repository.SearchPageRequestParams
import io.mike.foundation.repository.applySearchFilter
import io.mike.foundation.repository.toPageRequest
import io.mike.foundation.service.CrudService
import java.util.*

interface BookService : CrudService<Book, UUID, BookRepository, BookExtendedResponse> {

    fun find(params: SearchPageRequestParams): PageResponse<BookExtendedResponse> = find(params, setOf("title", "description"))
    fun findExtended(params: SearchPageRequestParams): PageResponse<BookExtendedResponse> =
            repository.findAll(applyBookSearchFilter(
                    params, setOf("title", "description"), setOf("name")
            ), params.toPageRequest()).toPageResponse()
}

File: typescript/src/contracts/books/BookService.ts
import { injectable } from "inversify";

import { CrudService } from "../../foundation/service";
import { NotImplementedError } from "../../foundation/core";
import { Book } from "./Book";
import { BookServiceProvider } from "./BookServiceProvider";

@injectable()
export class BookService extends CrudService<Book, string> implements BookServiceProvider {
    constructor() {
        super(Book);
    }

    public async find<Params>(params: Params): Promise<Book[]> {
        throw new NotImplementedError();
    }
}